src/mod_commands.erl

```diff
@@ -453,15 +500,53 @@ run_subscription(Type, CallerJid, OtherJid) ->
     ejabberd_router:route(CallerJid, OtherJid, Acc2),
     ok.
 
-set_subscription(Caller, Other, <<"connect">>) ->
+
+set_subscription(Caller, Other, Action) ->
+    case check_jids(Caller, Other) of
+        ok ->
+            case Action of
+                A when A == <<"connect">>; A == <<"disconnect">> ->
+                    do_set_subscription(Caller, Other, Action);
```

Comment:
This is an example of very redundant stringprepping. The `case` clause does `jid:from_binary/1` on each binary and then drops the built result. `do_set_subscription/3 `will then do twice `add_contact`, which each does three more `jid:from_binary/1`. That's already, over two binaries, eight `jid:from_binary/1`. But then `do_set_subscription/3` will call four times `subscription/3`, which in different combinations does another `check_jids` and some other `jid:from_binary/1`, amounting to four `jid:from_binary/1` for each `subscription/3`, of which there are four, so, 16 more `jid:from_binary/1`.

So, `set_subscription/3` with two binaries, will do 24 times `jid:from_binary/1`, twelve times for each 🤯🤣

I think two single `jid:from_binary/1` can be done at the very beginning, stop the call-stack if any is an error, and then run everything else passing the `#jid{}` record appropriately.